src/libraries/System.Net.Http/src/System/Net/Http/HttpClient.cs

```diff
@@ -506,63 +509,60 @@ public Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, HttpCompl
         public Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, HttpCompletionOption completionOption,
             CancellationToken cancellationToken)
         {
+            // Called outside of async state machine to propagate certain exception even without awaiting the returned task.
+            CheckRequestBeforeSend(request);
+
             return SendAsyncCore(request, completionOption, async: true, cancellationToken).AsTask();
         }
 
-        private ValueTask<HttpResponseMessage> SendAsyncCore(HttpRequestMessage request, HttpCompletionOption completionOption,
-            bool async, CancellationToken cancellationToken)
+        private void CheckRequestBeforeSend(HttpRequestMessage request)
         {
             if (request == null)
             {
                 throw new ArgumentNullException(nameof(request));
             }
             CheckDisposed();
-            CheckRequestMessage(request);
 
             SetOperationStarted();
-            PrepareRequestMessage(request);
+
             // PrepareRequestMessage will resolve the request address against the base address.
+            PrepareRequestMessage(request);
+        }
 
+        private async ValueTask<HttpResponseMessage> SendAsyncCore(HttpRequestMessage request, HttpCompletionOption completionOption,
+            bool async, CancellationToken cancellationToken)
+        {
             // Combines given cancellationToken with the global one and the timeout.
             CancellationTokenSource cts = PrepareCancellationTokenSource(cancellationToken, out bool disposeCts, out long timeoutTime);
 
-            // Initiate the send.
-            ValueTask<HttpResponseMessage> responseTask;
-            try
-            {
-                responseTask = async ?
-                    new ValueTask<HttpResponseMessage>(base.SendAsync(request, cts.Token)) :
-                    new ValueTask<HttpResponseMessage>(base.Send(request, cts.Token));
-            }
-            catch (Exception e)
-            {
-                HandleFinishSendCleanup(cts, disposeCts);
+            bool buffered = completionOption == HttpCompletionOption.ResponseContentRead &&
+                            !string.Equals(request.Method.Method, "HEAD", StringComparison.OrdinalIgnoreCase);
 
-                if (e is OperationCanceledException operationException && TimeoutFired(cancellationToken, timeoutTime))
+            bool telemetryStarted = false;
+            if (HttpTelemetry.Log.IsEnabled())
+            {
+                // [ActiveIssue("https://github.com/dotnet/runtime/issues/40896")]
+                if (request.Version.Major < 3 && request.RequestUri != null)
                 {
-                    throw CreateTimeoutException(operationException);
+                    HttpTelemetry.Log.RequestStart(
```

Comment:
So now you only get the telemetry if you use HtrpClient, but not if you use HttpMessageInvoker?

Reply:
Yes. Though in the previous state of things, you got the telemetry only with `SocketsHandler` and not with any custom handler...

However, we can introduce child events to cover `HttpMessageInvoker` etc. This event pair should be the main, all encompassing one. And in case a user wants to use the invoker directly, we should provide them with examples how to plug in their own telemetry to achieve the same.

I've some thoughts about this and I'm trying to put it all in writing in https://github.com/dotnet/runtime/issues/40776 (not yet finished though).